### ISSUE TITLE
No Map after Update to NodeJS14 / JSController 3.3

### ISSUE THREAD

**Issue by kschaumann** [iobroker-community-adapters/ioBroker.mihome-vacuum#397]
**Describe the bug**  
As also found for Update of NodeJS to 12 #119 , I have an equal problem after Update to 14. Or maybe after Update JSController to 3.3

As in #119 descriped, I have tried to: 
/opt/iobroker/node_modules/iobroker.mihome-vacuum $ npm i
and:
/opt/iobroker $ iobroker rebuild mihome-vacuum --install

I always restarted the adapter after this, no Errors are provided during this commands, except: 

`npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/gulp/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm"})
`

**To Reproduce**  
Map is not loaded/updated after roborock starts, any other value is updated periodically. 

**Expected behavior**  
The Map should update

**Screenshots & Logfiles**  
Relevant DebugLog: 

mihome-vacuum.0 | 2021-09-10 09:28:55.167 | debug | CanvasMap not loaded
mihome-vacuum.0 | 2021-09-10 09:28:55.166 | debug | update_Map got new time:1631258935
mihome-vacuum.0 | 2021-09-10 09:28:55.166 | debug | update_Map got new expires:1631260735
mihome-vacuum.0 | 2021-09-10 09:28:55.165 | debug | update_Map got new url:https://awsde0.fds.api.xiaomi.com/robomap/robomap/11UUUUU13/11?Expires=1631260735000&GalaxyAccessKeyId=5271XXXXX445&Signature=rMsYYYZZZXXXyLYkl8=
mihome-vacuum.0 | 2021-09-10 09:28:55.103 | debug | update_Map need new mapurl
mihome-vacuum.0 | 2021-09-10 09:28:55.102 | debug | update_Map Mimap enabled
mihome-vacuum.0 | 2021-09-10 09:28:55.101 | debug | Mappointer_updated
mihome-vacuum.0 | 2021-09-10 09:28:55.100 | debug | MIIO RECIVE: {"result":["robomap%2F118226113%2F11"],"id":54}
mihome-vacuum.0 | 2021-09-10 09:28:55.081 | debug | Message= {"id":54,"method":"get_map_v1"}
mihome-vacuum.0 | 2021-09-10 09:28:53.080 | debug | CanvasMap not loaded
mihome-vacuum.0 | 2021-09-10 09:28:53.079 | debug | update_Map got new time:1631258933
mihome-vacuum.0 | 2021-09-10 09:28:53.079 | debug | update_Map got new expires:1631260733
mihome-vacuum.0 | 2021-09-10 09:28:53.078 | debug | update_Map got new url:https://awsde0.fds.api.xiaomi.com/robomap/robomap/11UUUUU13/11?Expires=1631260735000&GalaxyAccessKeyId=5271XXXXX445&Signature=rMsYYYZZZXXXyLYkl8=

**Versions:**

-   Adapter version: v3.2.2
-   JS-Controller version: 3.3.15
-   Node version: 14.17.6
-   vacuum cleaner: S5

**Comment by Apollon77** [comment]
try executing "npm install" in the folder /opt/iobroker/node_modules/iobroker.mihome-vacuum

**Comment by kschaumann** [comment]
/opt/iobroker/node_modules/iobroker.mihome-vacuum $ npm install
`npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.2 (node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules/gulp/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm"})

removed 1 package and audited 657 packages in 10.66s

45 packages are looking for funding
  run npm fund for details

found 3 vulnerabilities (1 low, 2 moderate)
  run npm audit fix to fix them, or npm audit for details`

After Restart of Adapter, no changes: 


mihome-vacuum.0 | 2021-09-10 10:53:15.723 | debug | CanvasMap not loaded
mihome-vacuum.0 | 2021-09-10 10:53:15.722 | debug | update_Map got new time:1631263995
mihome-vacuum.0 | 2021-09-10 10:53:15.722 | debug | update_Map got new expires:1631265795
mihome-vacuum.0 | 2021-09-10 10:53:15.722 | debug | update_Map got new url:https://awsde0.fds.api.xiaomi.com/robomap/robomap/118226113/12?Expires=1631XXXX00&GalaxyAccessKeyId=5271733786445&Signature=HbgEDLR5XXXXXHYvl3YnIk=
mihome-vacuum.0 | 2021-09-10 10:53:15.666 | debug | MIIO RECIVE: {"result":[{"enable":1,"current_integral":450,"current_high":500,"current_low":400,"stall_time":10}],"id":132}
mihome-vacuum.0 | 2021-09-10 10:53:15.657 | debug | Message= {"id":132,"method":"get_carpet_mode"}
mihome-vacuum.0 | 2021-09-10 10:53:15.655 | debug | update_Map need new mapurl
mihome-vacuum.0 | 2021-09-10 10:53:15.655 | debug | update_Map Mimap enabled
mihome-vacuum.0 | 2021-09-10 10:53:15.655 | debug | Mappointer_updated
mihome-vacuum.0 | 2021-09-10 10:53:15.654 | debug | MIIO RECIVE: {"result":["robomap%2F118226113%2F12"],"id":131}
mihome-vacuum.0 | 2021-09-10 10:53:15.644 | debug | Message= {"id":131,"method":"get_map_v1"}
mihome-vacuum.0 | 2021-09-10 10:53:15.643 | debug | MIIO RECIVE: {"result":[57],"id":130}
mihome-vacuum.0 | 2021-09-10 10:53:15.619 | debug | Message= {"id":130,"method":"get_sound_volume"}
mihome-vacuum.0 | 2021-09-10 10:53:15.555 | debug | MIIO RECIVE: {"result":{"ssid":"MY_WLAN","ip":"192.168.178.ZZ","mac":"40:31:3c:ae:a5:46","bssid":"II:ff:4d:9d:e2:76","rssi":-61},"id":129}
mihome-vacuum.0 | 2021-09-10 10:53:15.529 | debug | Message= {"id":129,"method":"get_network_info"}
mihome-vacuum.0 | 2021-09-10 10:53:15.528 | debug | setGetStatus {"msg_ver":3,"msg_seq":1432,"state":8,"battery":40,"clean_time":6032,"clean_area":103812500,"error_code":0,"map_present":true,"in_cleaning":false,"in_returning":0,"in_fresh_state":1,"lab_status":1,"water_box_status":0,"fan_power":102,"dnd_enabled":false,"map_status":3,"lock_status":0,"error_text":"No error"}

**Comment by Apollon77** [comment]
then try a "npm rebuild" in /opt/iobroker

BACKUP vorher!!

**Comment by Apollon77** [comment]
@MeisterTR it is good that the adapter cactches the error that canvas module can not be leaded, but if it would expose the  error message (in this case would clearly state that canvas module needs a rebuild because of the nodejs upgrade also including the path) and could allow js-controller to try an auto fix

**Comment by kschaumann** [comment]
/opt/iobroker $ sudo npm rebuild

> iobroker@2.0.3 install /opt/iobroker/node_modules/iobroker
> node lib/checkVersions.js

NPM version: 6.14.15

> iobroker@2.0.3 postinstall /opt/iobroker/node_modules/iobroker
> node lib/install.js

lib/install.js is not being run as part of an installation - skipping...

> @serialport/bindings@9.2.1 install /opt/iobroker/node_modules/@serialport/bindings
> prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild

prebuild-install WARN install EACCES: permission denied, access '/root/.npm'
gyp ERR! clean error 
gyp ERR! stack Error: EACCES: permission denied, rmdir 'build'
gyp ERR! System Linux 5.10.60-v7l+
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /opt/iobroker/node_modules/@serialport/bindings
gyp ERR! node -v v14.17.6
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @serialport/bindings@9.2.1 install: `prebuild-install --tag-prefix @serialport/bindings@ || node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @serialport/bindings@9.2.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-09-10T09_10_52_770Z-debug.log


After iobroker start und update the Map: 

mihome-vacuum.0 | 2021-09-10 11:17:51.790 | debug | CanvasMap not loaded
mihome-vacuum.0 | 2021-09-10 11:17:51.789 | debug | update_Map got new time:1631265471
mihome-vacuum.0 | 2021-09-10 11:17:51.789 | debug | update_Map got new expires:1631267271
mihome-vacuum.0 | 2021-09-10 11:17:51.787 | debug | update_Map got new url:https://awsde0.fds.api.xiaomi.com/robomap/robomap/118226113/13?Expires=1631267271000&GalaxyAccessKeyId=5271733786445&Signature=ybTYYYYrUE/ynaWlYYYYYYPLeRoE=
mihome-vacuum.0 | 2021-09-10 11:17:51.726 | debug | MIIO RECIVE: {"result":[{"enable":1,"current_integral":450,"current_high":500,"current_low":400,"stall_time":10}],"id":90}
mihome-vacuum.0 | 2021-09-10 11:17:51.713 | debug | Message= {"id":90,"method":"get_carpet_mode"}
mihome-vacuum.0 | 2021-09-10 11:17:51.712 | debug | update_Map need new mapurl
mihome-vacuum.0 | 2021-09-10 11:17:51.711 | debug | update_Map Mimap enabled
mihome-vacuum.0 | 2021-09-10 11:17:51.711 | debug | Mappointer_updated

**Comment by kschaumann** [comment]
Additial Version Infos: 
/opt/iobroker $ node -v
v14.17.6
pi@raspberrypi:/opt/iobroker $ nodejs -v
v14.17.6
pi@raspberrypi:/opt/iobroker $ npm -v
6.14.15
pi@raspberrypi:/opt/iobroker $ iobroker -v
3.3.15

**Comment by kschaumann** [comment]
I've fixed it!
After checking the needed libs with: 
`sudo apt-get install build-essential libcairo2-dev libpango1.0-dev libjpeg-dev libgif-dev librsvg2-dev`

Which results, that nothing is missing, I tried: 
`sudo npm install canvas --unsafe-perm=true`

Then I restart the Adapter via the Admin-GUI and some times later I get the Map updated!

YES!

A little bit strange, cause everything was installed before the Update of Node(JS) and the JS-Controller.
